### PR TITLE
8342902: Deduplication of acquire calls in BindingSpecializer causes escape-analyisis failure

### PR DESCRIPTION
**Pull request by mcimadamore**
This PR fixes an issue where passing many by-reference parameters to downcall results in escape analysis failures.
The problem is that, as the parameters grow, the generated code in the trampoline stub we generate also grows.
When it reaches a certain threshold, it becomes too big, and it is no longer inlined in the caller.
When this happens, allocations for by-reference parameters (e.g. a segment constructed from `MemorySegment::ofAddress`) can no longer be eliminated.

The solution is two-fold. First, we annotate the generated trampoline with `@ForceInline`. After all, it is rather critical, to guarantee optimal performance, that this code can be always inlined.
Second, to keep the size of the generated code under control, we also put a limit on the max number of comparisons that are generated in order to "deduplicate" scope acquire/release calls.
The deduplication logic is a bit finicky -- it was put in place because, when confined and shared are passed by-ref, we need to prevent them from being closed in the middle of a native call.
So, we save all the seen scopes in a bunch of locals, and then we compare each new scope with _all_ the previous cached locals, and skip acquire if we can.

While this strategy work it does not scale when there are many by-ref parameters - as a by-ref parameter N will need N - 1 comparisons - which means a quadratic number of comparisons is generated.
This is fixed in this PR by putting a lid on the maximum number of comparisons that are generated. We also make the comparisons a bit smarter, by always skipping the very first by-ref argument -- the downcall address.
It is in fact very common for the downcall address to be in a different scope than that of the other by-ref arguments anyway.

A nice property of the new logic is that by configuring the max number of comparisons we can effectively select between different strategies:
* max = 0, means no dedup
* max = 1, means one-element cache
* max = N, means full dedup (like before)

Thanks to Ioannis (@spasi) for the report and the benchmark. I've beefed the benchmark up by adding a case for 10 arguments, and also adding support for critical downcalls, so we could also test passing by-ref heap segments. Benchmark result will be provided in a separate comment.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342902](https://bugs.openjdk.org/browse/JDK-8342902): Deduplication of acquire calls in BindingSpecializer causes escape-analyisis failure (**Bug** - P3)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21706/head:pull/21706` \
`$ git checkout pull/21706`

Update a local copy of the PR: \
`$ git checkout pull/21706` \
`$ git pull https://git.openjdk.org/jdk.git pull/21706/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21706`

View PR using the GUI difftool: \
`$ git pr show -t 21706`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21706.diff">https://git.openjdk.org/jdk/pull/21706.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21706#issuecomment-2437227852)

### CODE REVIEW

**Comment by mcimadamore** [comment]
This are the results of running the new benchmark on my workstation:

```
Benchmark                                               (kind)  Mode  Cnt   Score    Error   Units
CallByRefHighArity.noop_params0                       CONFINED  avgt    3   2.714 ±  0.444   ns/op
CallByRefHighArity.noop_params0:·gc.alloc.rate        CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params0:·gc.alloc.rate.norm   CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params0:·gc.count             CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params0                         SHARED  avgt    3   2.795 ±  0.045   ns/op
CallByRefHighArity.noop_params0:·gc.alloc.rate          SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params0:·gc.alloc.rate.norm     SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params0:·gc.count               SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params0                         GLOBAL  avgt    3   2.762 ±  0.321   ns/op
CallByRefHighArity.noop_params0:·gc.alloc.rate          GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params0:·gc.alloc.rate.norm     GLOBAL  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params0:·gc.count               GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params0                           HEAP  avgt    3   2.775 ±  0.330   ns/op
CallByRefHighArity.noop_params0:·gc.alloc.rate            HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params0:·gc.alloc.rate.norm       HEAP  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params0:·gc.count                 HEAP  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params1                       CONFINED  avgt    3   4.667 ±  0.207   ns/op
CallByRefHighArity.noop_params1:·gc.alloc.rate        CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params1:·gc.alloc.rate.norm   CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params1:·gc.count             CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params1                         SHARED  avgt    3   7.942 ±  4.004   ns/op
CallByRefHighArity.noop_params1:·gc.alloc.rate          SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params1:·gc.alloc.rate.norm     SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params1:·gc.count               SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params1                         GLOBAL  avgt    3   2.922 ±  0.397   ns/op
CallByRefHighArity.noop_params1:·gc.alloc.rate          GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params1:·gc.alloc.rate.norm     GLOBAL  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params1:·gc.count               GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params1                           HEAP  avgt    3   3.668 ±  0.255   ns/op
CallByRefHighArity.noop_params1:·gc.alloc.rate            HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params1:·gc.alloc.rate.norm       HEAP  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params1:·gc.count                 HEAP  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params2                       CONFINED  avgt    3   4.443 ±  0.383   ns/op
CallByRefHighArity.noop_params2:·gc.alloc.rate        CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params2:·gc.alloc.rate.norm   CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params2:·gc.count             CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params2                         SHARED  avgt    3   7.938 ±  0.644   ns/op
CallByRefHighArity.noop_params2:·gc.alloc.rate          SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params2:·gc.alloc.rate.norm     SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params2:·gc.count               SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params2                         GLOBAL  avgt    3   2.923 ±  0.446   ns/op
CallByRefHighArity.noop_params2:·gc.alloc.rate          GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params2:·gc.alloc.rate.norm     GLOBAL  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params2:·gc.count               GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params2                           HEAP  avgt    3   3.791 ±  0.103   ns/op
CallByRefHighArity.noop_params2:·gc.alloc.rate            HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params2:·gc.alloc.rate.norm       HEAP  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params2:·gc.count                 HEAP  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params3                       CONFINED  avgt    3   4.793 ±  2.755   ns/op
CallByRefHighArity.noop_params3:·gc.alloc.rate        CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params3:·gc.alloc.rate.norm   CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params3:·gc.count             CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params3                         SHARED  avgt    3   8.842 ±  0.786   ns/op
CallByRefHighArity.noop_params3:·gc.alloc.rate          SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params3:·gc.alloc.rate.norm     SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params3:·gc.count               SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params3                         GLOBAL  avgt    3   2.726 ±  0.114   ns/op
CallByRefHighArity.noop_params3:·gc.alloc.rate          GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params3:·gc.alloc.rate.norm     GLOBAL  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params3:·gc.count               GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params3                           HEAP  avgt    3   4.155 ±  0.581   ns/op
CallByRefHighArity.noop_params3:·gc.alloc.rate            HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params3:·gc.alloc.rate.norm       HEAP  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params3:·gc.count                 HEAP  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params4                       CONFINED  avgt    3   5.008 ±  0.361   ns/op
CallByRefHighArity.noop_params4:·gc.alloc.rate        CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params4:·gc.alloc.rate.norm   CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params4:·gc.count             CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params4                         SHARED  avgt    3   9.612 ±  0.722   ns/op
CallByRefHighArity.noop_params4:·gc.alloc.rate          SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params4:·gc.alloc.rate.norm     SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params4:·gc.count               SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params4                         GLOBAL  avgt    3   3.076 ±  0.250   ns/op
CallByRefHighArity.noop_params4:·gc.alloc.rate          GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params4:·gc.alloc.rate.norm     GLOBAL  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params4:·gc.count               GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params4                           HEAP  avgt    3   4.520 ±  0.275   ns/op
CallByRefHighArity.noop_params4:·gc.alloc.rate            HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params4:·gc.alloc.rate.norm       HEAP  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params4:·gc.count                 HEAP  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params5                       CONFINED  avgt    3   5.429 ±  2.031   ns/op
CallByRefHighArity.noop_params5:·gc.alloc.rate        CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params5:·gc.alloc.rate.norm   CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params5:·gc.count             CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params5                         SHARED  avgt    3   9.716 ±  2.919   ns/op
CallByRefHighArity.noop_params5:·gc.alloc.rate          SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params5:·gc.alloc.rate.norm     SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params5:·gc.count               SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params5                         GLOBAL  avgt    3   2.954 ±  0.035   ns/op
CallByRefHighArity.noop_params5:·gc.alloc.rate          GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params5:·gc.alloc.rate.norm     GLOBAL  avgt    3  ≈ 10⁻⁶             B/op
CallByRefHighArity.noop_params5:·gc.count               GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params5                           HEAP  avgt    3   5.001 ±  1.058   ns/op
CallByRefHighArity.noop_params5:·gc.alloc.rate            HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params5:·gc.alloc.rate.norm       HEAP  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params5:·gc.count                 HEAP  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params10                      CONFINED  avgt    3   8.532 ±  1.880   ns/op
CallByRefHighArity.noop_params10:·gc.alloc.rate       CONFINED  avgt    3   0.001 ±  0.015  MB/sec
CallByRefHighArity.noop_params10:·gc.alloc.rate.norm  CONFINED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params10:·gc.count            CONFINED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params10                        SHARED  avgt    3  12.237 ±  2.431   ns/op
CallByRefHighArity.noop_params10:·gc.alloc.rate         SHARED  avgt    3   0.001 ±  0.026  MB/sec
CallByRefHighArity.noop_params10:·gc.alloc.rate.norm    SHARED  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params10:·gc.count              SHARED  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params10                        GLOBAL  avgt    3   7.216 ±  0.846   ns/op
CallByRefHighArity.noop_params10:·gc.alloc.rate         GLOBAL  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params10:·gc.alloc.rate.norm    GLOBAL  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params10:·gc.count              GLOBAL  avgt    3     ≈ 0           counts
CallByRefHighArity.noop_params10                          HEAP  avgt    3   8.961 ±  0.667   ns/op
CallByRefHighArity.noop_params10:·gc.alloc.rate           HEAP  avgt    3  ≈ 10⁻³           MB/sec
CallByRefHighArity.noop_params10:·gc.alloc.rate.norm      HEAP  avgt    3  ≈ 10⁻⁵             B/op
CallByRefHighArity.noop_params10:·gc.count                HEAP  avgt    3     ≈ 0           counts
```

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java [hunk 1]
```diff
@@ -77,6 +81,8 @@ public class BindingSpecializer {
         = GetPropertyAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.DUMP_CLASSES_DIR");
     private static final boolean PERFORM_VERIFICATION
         = GetBooleanAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.PERFORM_VERIFICATION");
+    private static final int SCOPE_DEDUP_DEPTH
+            = GetIntegerAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.SCOPE_DEDUP_DEPTH", 2);
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:85]: A two-comparison logic seems like a good compromise. Note this doesn't really mean a two-element cache - it is still possible to have e.g. the first two arguments with the same scope and then a third argumemt with another scope - in which case the third argument won't be deduped. This is a trade-off - to keep the generated code simple we need to have a 1-1 mapping between by-ref parameters and locals.

File: test/micro/org/openjdk/bench/java/lang/foreign/libCallByRefHighArity.c [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include "export.h"
+
+EXPORT void noop_params0() {}
+EXPORT void noop_params1(void *param0) {}
+EXPORT void noop_params2(void *param0, void *param1) {}
+EXPORT void noop_params3(void *param0, void *param1, void *param2) {}
+EXPORT void noop_params4(void *param0, void *param1, void *param2, void *param3) {}
+EXPORT void noop_params5(int param0, int param1, void *param2, void *param3, void *param4) {}
+EXPORT void noop_params10(int param0, int param1, void *param2, void *param3, void *param4,
+                          int param5, int param6, void *param7, void *param8, void *param9) {}
```
Comment (JornVernee) [inline test/micro/org/openjdk/bench/java/lang/foreign/libCallByRefHighArity.c:33]: Some of these are `int` which mismatches the `ADDRESS` layout used to create the function descriptor.

**Review by JornVernee — APPROVED** [review]
Looks great, aside from the native lib issue (inline)

File: test/micro/org/openjdk/bench/java/lang/foreign/libCallByRefHighArity.c (hunk 2 above)
Comment (mcimadamore) [reply test/micro/org/openjdk/bench/java/lang/foreign/libCallByRefHighArity.c:33]: ugh - true! Will replace with `void*`

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**